File: full-stack-spec-driven-dev/scripts/generate-index.py
#!/usr/bin/env python3
"""
Generate specs/INDEX.md from all spec files.

Usage: python generate-index.py --specs-dir specs/
"""

import argparse
import re
from pathlib import Path
from datetime import datetime
from collections import defaultdict


def parse_frontmatter(content: str) -> dict:
    """Extract YAML frontmatter from markdown."""
    match = re.match(r'^---\s*\n(.*?)\n---', content, re.DOTALL)
    if not match:
        return {}

    frontmatter = {}
    for line in match.group(1).split('\n'):
        if ':' in line:
            key, value = line.split(':', 1)
            frontmatter[key.strip()] = value.strip()
    return frontmatter


def generate_index(specs_dir: Path) -> str:
    """Generate INDEX.md content."""
    # Find all spec files
    specs = [p for p in specs_dir.rglob('*.md')
             if p.name not in ('INDEX.md', 'SNAPSHOT.md', 'glossary.md')]

    # Parse and categorize
    by_status = defaultdict(list)
    for spec_path in specs:
        content = spec_path.read_text()
        fm = parse_frontmatter(content)

        status = fm.get('status', 'active')
        title = fm.get('title', spec_path.stem)
        change_type = fm.get('type', 'feature')
        domain = fm.get('domain', 'Unknown')
        issue = fm.get('issue', '')
        created = fm.get('created', '')

        # Make path relative to specs dir
        rel_path = spec_path.relative_to(specs_dir)

        by_status[status].append({
            'title': title,
            'type': change_type,
            'path': str(rel_path),
            'domain': domain,
            'issue': issue,
            'created': created,
        })

    # Count totals
    total = len(specs)
    active = len(by_status['active'])
    deprecated = len(by_status['deprecated'])
    archived = len(by_status['archived'])

    # Generate markdown
    lines = [
        "# Spec Index",
        "",
        f"Last updated: {datetime.now().strftime('%Y-%m-%d')}",
        "",
        f"Total: {total} specs (Active: {active}, Deprecated: {deprecated}, Archived: {archived})",
        "",
    ]

    # Active specs
    lines.extend([
        "## Active Changes",
        "",
        "| Change | Type | Spec | Domain | Issue | Since |",
        "|--------|------|------|--------|-------|-------|",
    ])

    if by_status['active']:
        for spec in sorted(by_status['active'], key=lambda x: x['created']):
            issue_link = f"[{spec['issue']}](#)" if spec['issue'] else ""
            lines.append(
                f"| {spec['title']} | {spec['type']} | [{spec['path']}]({spec['path']}) | {spec['domain']} | {issue_link} | {spec['created']} |"
            )
    else:
        lines.append("| *No active changes yet* | | | | | |")

    lines.append("")

    # Deprecated specs
    lines.extend([
        "## Deprecated",
        "",
    ])

    if by_status['deprecated']:
        lines.append("| Change | Type | Spec | Domain | Issue | Deprecated |")
        lines.append("|--------|------|------|--------|-------|------------|")
        for spec in sorted(by_status['deprecated'], key=lambda x: x['created']):
            issue_link = f"[{spec['issue']}](#)" if spec['issue'] else ""
            lines.append(
                f"| {spec['title']} | {spec['type']} | [{spec['path']}]({spec['path']}) | {spec['domain']} | {issue_link} | {spec['created']} |"
            )
    else:
        lines.append("*None*")

    lines.append("")

    # Archived specs
    lines.extend([
        "## Archived",
        "",
    ])

    if by_status['archived']:
        lines.append("| Change | Type | Spec | Domain | Issue | Archived |")
        lines.append("|--------|------|------|--------|-------|----------|")
        for spec in sorted(by_status['archived'], key=lambda x: x['created']):
            issue_link = f"[{spec['issue']}](#)" if spec['issue'] else ""
            lines.append(
                f"| {spec['title']} | {spec['type']} | [{spec['path']}]({spec['path']}) | {spec['domain']} | {issue_link} | {spec['created']} |"
            )
    else:
        lines.append("*None*")

    return '\n'.join(lines) + '\n'


def main():
    parser = argparse.ArgumentParser(description='Generate spec index')
    parser.add_argument('--specs-dir', default='specs/', help='Specs directory')
    args = parser.parse_args()

    specs_dir = Path(args.specs_dir)
    if not specs_dir.exists():
        print(f"Error: Specs directory not found: {specs_dir}")
        return 1

    index_content = generate_index(specs_dir)
    index_path = specs_dir / 'INDEX.md'
    index_path.write_text(index_content)

    print(f"✓ Generated {index_path}")
    return 0


if __name__ == '__main__':
    import sys
    sys.exit(main())

File: plugins/sdd/scripts/generate-index.ts
#!/usr/bin/env npx ts-node --esm
/**
 * Generate specs/INDEX.md from all spec files.
 *
 * Usage: npx ts-node --esm generate-index.ts --specs-dir specs/
 */

import * as fs from 'node:fs/promises';
import * as path from 'node:path';
import { findSpecFiles, directoryExists } from './lib/spec-utils.js';

interface SpecEntry {
  readonly title: string;
  readonly type: string;
  readonly path: string;
  readonly domain: string;
  readonly issue: string;
  readonly created: string;
}

/**
 * Generate INDEX.md content.
 */
const generateIndex = async (specsDir: string): Promise<string> => {
  const specs = await findSpecFiles(specsDir);

  // Categorize by status
  const byStatus: Record<string, SpecEntry[]> = {
    active: [],
    deprecated: [],
    archived: [],
  };

  for (const spec of specs) {
    const fm = spec.frontmatter ?? {};
    const status = fm['status'] ?? 'active';
    const entry: SpecEntry = {
      title: fm['title'] ?? path.basename(spec.path, '.md'),
      type: fm['type'] ?? 'feature',
      path: spec.relativePath,
      domain: fm['domain'] ?? 'Unknown',
      issue: fm['issue'] ?? '',
      created: fm['created'] ?? '',
    };

    if (status in byStatus) {
      byStatus[status]?.push(entry);
    }
  }

  // Count totals
  const total = specs.length;
  const active = byStatus['active']?.length ?? 0;
  const deprecated = byStatus['deprecated']?.length ?? 0;
  const archived = byStatus['archived']?.length ?? 0;

  const today = new Date().toISOString().split('T')[0];

  // Generate markdown
  const lines: string[] = [
    '# Spec Index',
    '',
    `Last updated: ${today}`,
    '',
    `Total: ${total} specs (Active: ${active}, Deprecated: ${deprecated}, Archived: ${archived})`,
    '',
    '## Active Changes',
    '',
    '| Change | Type | Spec | Domain | Issue | Since |',
    '|--------|------|------|--------|-------|-------|',
  ];

  const activeSpecs = byStatus['active'] ?? [];
  if (activeSpecs.length > 0) {
    const sorted = [...activeSpecs].sort((a, b) => a.created.localeCompare(b.created));
    for (const spec of sorted) {
      const issueLink = spec.issue ? `[${spec.issue}](#)` : '';
      lines.push(
        `| ${spec.title} | ${spec.type} | [${spec.path}](${spec.path}) | ${spec.domain} | ${issueLink} | ${spec.created} |`
      );
    }
  } else {
    lines.push('| *No active changes yet* | | | | | |');
  }

  lines.push('');
  lines.push('## Deprecated');
  lines.push('');

  const deprecatedSpecs = byStatus['deprecated'] ?? [];
  if (deprecatedSpecs.length > 0) {
    lines.push('| Change | Type | Spec | Domain | Issue | Deprecated |');
    lines.push('|--------|------|------|--------|-------|------------|');
    const sorted = [...deprecatedSpecs].sort((a, b) => a.created.localeCompare(b.created));
    for (const spec of sorted) {
      const issueLink = spec.issue ? `[${spec.issue}](#)` : '';
      lines.push(
        `| ${spec.title} | ${spec.type} | [${spec.path}](${spec.path}) | ${spec.domain} | ${issueLink} | ${spec.created} |`
      );
    }
  } else {
    lines.push('*None*');
  }

  lines.push('');
  lines.push('## Archived');
  lines.push('');

  const archivedSpecs = byStatus['archived'] ?? [];
  if (archivedSpecs.length > 0) {
    lines.push('| Change | Type | Spec | Domain | Issue | Archived |');
    lines.push('|--------|------|------|--------|-------|----------|');
    const sorted = [...archivedSpecs].sort((a, b) => a.created.localeCompare(b.created));
    for (const spec of sorted) {
      const issueLink = spec.issue ? `[${spec.issue}](#)` : '';
      lines.push(
        `| ${spec.title} | ${spec.type} | [${spec.path}](${spec.path}) | ${spec.domain} | ${issueLink} | ${spec.created} |`
      );
    }
  } else {
    lines.push('*None*');
  }

  return lines.join('\n') + '\n';
};

/**
 * Parse command line arguments.
 */
const parseArgs = (args: readonly string[]): { specsDir: string } => {
  let specsDir = 'specs/';

  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (arg === '--specs-dir') {
      specsDir = args[i + 1] ?? 'specs/';
      i++;
    }
  }

  return { specsDir };
};

const main = async (): Promise<number> => {
  const args = parseArgs(process.argv.slice(2));

  if (!(await directoryExists(args.specsDir))) {
    console.error(`Error: Specs directory not found: ${args.specsDir}`);
    return 1;
  }

  const indexContent = await generateIndex(args.specsDir);
  const indexPath = path.join(args.specsDir, 'INDEX.md');
  await fs.writeFile(indexPath, indexContent);

  console.log(`✓ Generated ${indexPath}`);
  return 0;
};

main()
  .then(process.exit)
  .catch((err) => {
    console.error(err);
    process.exit(1);
  });
